`makeClassyPrisms` double definitions

Something is rotten in Denmark.

```
newtype SDLException = SDLException String deriving Show
makeClassyPrisms ''SDLException
```

```
src/Engine/SDL.hs:20:1: error:
    Multiple declarations of ‘_SDLException’
    Declared at: src/Engine/SDL.hs:20:1
                 src/Engine/SDL.hs:20:1
   |
```

I don't use classy prisms very often, but this seems broken, as this is almost a textbook usecase.

--- a/CHANGELOG.markdown
+++ b/CHANGELOG.markdown
@@ -7,6 +7,11 @@
 * Weaken `holeInOne`'s `Category p` constraint to `Comonad (Corep p)`.
 * Generalize the type of `GHC.Generics.Lens.generic1` from
   `Iso' (f a) (Rep1 f a)` to `Iso (f a) (f b) (Rep1 f a) (Rep1 f b)`.
+* `makeClassyPrisms` now supports generating prisms for data types that share
+  a name with one of its constructors. In such a scenario, the name of the
+  classy prism for the data type will be prefixed with an extra `_` (for
+  prefix names) or `.` (for infix names) to disambiguate it from the prism
+  for the constructor.
 * Add additional bifunctor instances for `Swapped`.
 * New lenses `head1` and `last1`, to access the first/last elements of
   a `Traversable1` container.

--- a/src/Control/Lens/Internal/PrismTH.hs
+++ b/src/Control/Lens/Internal/PrismTH.hs
@@ -105,6 +105,31 @@ makePrisms = makePrisms' True
 -- Generate an "As" class of prisms. Names are selected by prefixing the constructor
 -- name with an underscore.  Constructors with multiple fields will
 -- construct Prisms to tuples of those fields.
+--
+-- In the event that the name of a data type is also the name of one of its
+-- constructors, the name of the 'Prism' generated for the data type will be
+-- prefixed with an extra @_@ (if the data type name is prefix) or @.@ (if the
+-- name is infix) to disambiguate it from the 'Prism' for the corresponding
+-- constructor. For example, this code:
+--
+-- @
+-- data Quux = Quux Int | Fred Bool
+-- makeClassyPrisms ''Quux
+-- @
+--
+-- will create:
+--
+-- @
+-- class AsQuux s where
+--   __Quux :: Prism' s Quux -- Data type prism
+--   _Quux :: Prism' s Int   -- Constructor prism
+--   _Fred :: Prism' s Bool
+--
+--   _Quux = __Quux . _Quux
+--   _Fred = __Quux . _Fred
+--
+-- instance AsQuux Quux
+-- @
 makeClassyPrisms :: Name {- ^ Type constructor name -} -> DecsQ
 makeClassyPrisms = makePrisms' False
 
@@ -157,8 +182,10 @@ makeConsPrisms t cons (Just typeName) =
     , makeClassyPrismInstance t className methodName cons
     ]
   where
-  className = mkName ("As" ++ nameBase typeName)
-  methodName = prismName typeName
+  typeNameBase = nameBase typeName
+  className = mkName ("As" ++ typeNameBase)
+  sameNameAsCon = any (\con -> nameBase (view nconName con) == typeNameBase) cons
+  methodName = prismName' sameNameAsCon typeName
 
 
 data OpticType = PrismType | ReviewType
@@ -485,11 +512,29 @@ normalizeCon info = NCon (D.constructorName info)
 -- | Compute a prism's name by prefixing an underscore for normal
 -- constructors and period for operators.
 prismName :: Name -> Name
-prismName n = case nameBase n of
-                [] -> error "prismName: empty name base?"
-                x:xs | isUpper x -> mkName ('_':x:xs)
-                     | otherwise -> mkName ('.':x:xs) -- operator
+prismName = prismName' False
 
+prismName' :: Bool -- ^ This is 'True' in the event that:
+                   --
+                   -- 1. We are generating the name of a classy prism for a
+                   --    data type, and
+                   -- 2. The data type shares a name with one of its
+                   --    constructors (e.g., @data A = A@).
+                   --
+                   -- In such a scenario, we take care not to generate the same
+                   -- prism name that the constructor receives (e.g., @_A@).
+                   -- For prefix names, we accomplish this by adding an extra
+                   -- underscore; for infix names, an extra dot.
+           -> Name -> Name
+prismName' sameNameAsCon n =
+  case nameBase n of
+    [] -> error "prismName: empty name base?"
+    nb@(x:_) | isUpper x -> mkName (prefix '_' nb)
+             | otherwise -> mkName (prefix '.' nb) -- operator
+  where
+    prefix :: Char -> String -> String
+    prefix char str | sameNameAsCon = char:char:str
+                    | otherwise     =      char:str
 
 -- | Quantify all the free variables in a type.
 close :: Type -> TypeQ

--- a/src/Control/Lens/Internal/PrismTH.hs
+++ b/src/Control/Lens/Internal/PrismTH.hs
@@ -514,18 +514,24 @@ normalizeCon info = NCon (D.constructorName info)
 prismName :: Name -> Name
 prismName = prismName' False
 
-prismName' :: Bool -- ^ This is 'True' in the event that:
-                   --
-                   -- 1. We are generating the name of a classy prism for a
-                   --    data type, and
-                   -- 2. The data type shares a name with one of its
-                   --    constructors (e.g., @data A = A@).
-                   --
-                   -- In such a scenario, we take care not to generate the same
-                   -- prism name that the constructor receives (e.g., @_A@).
-                   -- For prefix names, we accomplish this by adding an extra
-                   -- underscore; for infix names, an extra dot.
-           -> Name -> Name
+-- | Compute a prism's name with a special case for when the type
+-- constructor matches one of the value constructors.
+--
+-- The overlapping flag wil be 'True' in the event that:
+--
+-- 1. We are generating the name of a classy prism for a
+--    data type, and
+-- 2. The data type shares a name with one of its
+--    constructors (e.g., @data A = A@).
+--
+-- In such a scenario, we take care not to generate the same
+-- prism name that the constructor receives (e.g., @_A@).
+-- For prefix names, we accomplish this by adding an extra
+-- underscore; for infix names, an extra dot.
+prismName' ::
+  Bool {- ^ overlapping constructor -} ->
+  Name {- ^ type constructor        -} ->
+  Name {- ^ prism name              -}
 prismName' sameNameAsCon n =
   case nameBase n of
     [] -> error "prismName: empty name base?"

--- a/tests/templates.hs
+++ b/tests/templates.hs
@@ -445,5 +445,12 @@ checkAbbreviatedNamer = fieldAbbreviatedNamer
 data T866 = MkT866
 $(makeClassyPrisms ''T866)
 
+-- Ensure that `makeClassyPrisms` doesn't generate duplicate prism names for
+-- data types that share a name with one of its constructors (#865)
+data T865 = T865 | T865a | T865b T866
+$(makeClassyPrisms ''T865)
+instance AsT866 T865 where
+  _T866 = __T865 . _T866
+
 main :: IO ()
 main = putStrLn "test/templates.hs: ok"

